### PR TITLE
[MLOPS-596] Add Lakehouse Monitor Resource

### PR DESCRIPTION
**Pull request by aravind-segu**
## Changes
Adds Lakehouse Monitor Resource to the Terraform Provider

## Tests

- [x] `make test` run locally
- [x] relevant change in `docs/` folder
- [x] covered with integration tests in `internal/acceptance`
- [x] relevant acceptance tests are passing
- [x] using Go SDK

### CODE REVIEW

File: monitoring/resourse_lakehouse_monitor.go
Comment (alexott) [inline monitoring/resourse_lakehouse_monitor.go:22]: Why do we need this struct? Why can't we use `catalog.MonitorInfo` directly?

File: monitoring/resourse_lakehouse_monitor.go
Comment (alexott) [inline monitoring/resourse_lakehouse_monitor.go:43]: All of these should be marked as `computed` when customizing the schema...

File: monitoring/resourse_lakehouse_monitor.go
Comment (alexott) [inline monitoring/resourse_lakehouse_monitor.go:68]: These customizations aren't reflected in the Terraform schema because the `Schema` field is set to `readSchema`

File: monitoring/resourse_lakehouse_monitor.go
Comment (alexott) [inline monitoring/resourse_lakehouse_monitor.go:32]: If snapshot isn't used right now, we can remove it completely for now... This could be done in schema customization function as well

**Review by alexott — CHANGES_REQUESTED** [review]

File: monitoring/resourse_lakehouse_monitor.go
Comment (alexott) [reply monitoring/resourse_lakehouse_monitor.go:43]: The lack of `computed` leads to failure of the integration tests:

```
          ~ resource "databricks_lakehouse_monitor" "testMonitor" {
              - drift_metrics_table_name   = "sandbox$tblbheiblgbgc.things$tblbheiblgbgc.bar$tblbheiblgbgc_drift_metrics" -> null
                id                         = "sandbox$tblbheiblgbgc.things$tblbheiblgbgc.bar$tblbheiblgbgc"
              - profile_metrics_table_name = "sandbox$tblbheiblgbgc.things$tblbheiblgbgc.bar$tblbheiblgbgc_profile_metrics" -> null
              - status                     = "MONITOR_STATUS_PENDING" -> null
                # (3 unchanged attributes hidden)
        
                # (1 unchanged block hidden)
            }
        
        Plan: 0 to add, 1 to change, 0 to destroy.
```

File: monitoring/resourse_lakehouse_monitor.go
Comment (aravind-segu) [reply monitoring/resourse_lakehouse_monitor.go:22]: We need this struct because of the `Snapshot` field. Snapshot is of type `any` and it is not compatible with the schema. Therefore this custom struct uses an empty struct for snapshot, and then initializes the snapshot field later in the create and update below.

File: monitoring/resourse_lakehouse_monitor.go
Comment (aravind-segu) [reply monitoring/resourse_lakehouse_monitor.go:32]: Snapshot is being used. There are three types of monitors, Snapshot, Timeseries and Inference Log. For a snapshot there are no extra parameters, so the user just needs to instantiate the Snapshot field in order to create a snapshot monitor.

**Comment by alexott** [comment]
@aravind-segu integration tests are passing (at least on AWS)

File: monitoring/resourse_lakehouse_monitor.go
Comment (mgyucht) [reply monitoring/resourse_lakehouse_monitor.go:22]: Why is `Snapshot` `any`? What is its actual type? Is it mean to be `struct{}`, and this is a bug with the types in the Go SDK? Or can users actually pass any value for this field?

**Review by mgyucht — CHANGES_REQUESTED** [review]
Thank you for contributing this resource! We do need to make some adjustments to improve the long-term maintainability story, and they may necessitate deeper changes. If it isn't possible to do this right from the start, we need to prioritize the work needed for this to be properly supported.

File: monitoring/resourse_lakehouse_monitor.go
Comment (BoopBoopBeepBoop) [reply monitoring/resourse_lakehouse_monitor.go:97]: the ID of the monitor should not be changed

File: monitoring/resourse_lakehouse_monitor.go
Comment (alexott) [inline monitoring/resourse_lakehouse_monitor.go:91]: Why do we need this? `DataToStructPointer` deserialize things - you already have `FullName` annotated as `table_name`...

File: monitoring/resourse_lakehouse_monitor.go
Comment (alexott) [inline monitoring/resourse_lakehouse_monitor.go:32]: If `Snapshot` is optional, then it should be a pointer, not struct: 

```
Snapshot                 *Snapshot                                 `json:"snapshot,omitempty"`
```

File: monitoring/resourse_lakehouse_monitor.go
Comment (alexott) [inline monitoring/resourse_lakehouse_monitor.go:43]: This also should be a pointer

File: monitoring/resourse_lakehouse_monitor.go
Comment (mgyucht) [reply monitoring/resourse_lakehouse_monitor.go:32]: This is a bug in the Go SDK. The Go SDK currently treats empty structures and components that can take on any value in the same way. I am opening a PR to fix this here: https://github.com/databricks/databricks-sdk-go/pull/821

File: monitoring/resourse_lakehouse_monitor.go
Comment (aravind-segu) [reply monitoring/resourse_lakehouse_monitor.go]: MonitorInfoStatus is just a string enum. If I make it a pointer, the `StructToSchema` method expects it to be a struct.

File: monitoring/resourse_lakehouse_monitor.go
Comment (aravind-segu) [reply monitoring/resourse_lakehouse_monitor.go]: In the `catalog.CreateMonitor` struct we define the `FullName` with the json path `json:"-" url:"-"`. Therefore we have to set it manually as there is no json path that the schema can set it to.

**Comment by alexott** [comment]
Integration tests are failing - I fixed incorrect template (please pull changes), but even after that, it looks like `Update` method doesn't send the correct payload:

```
2024-02-24T11:02:07.276Z [DEBUG] databricks: PUT /api/2.1/unity-catalog/tables/sandbox$tdeaabhaalaji.things$tabikdcljiild.bar$tabikdcljiild_inference/monitor
> {
>   "inference_log": {
>     "granularities": [
>       "1 hour"
>     ],
>     "model_id_col": "model_id",
>     "prediction_col": "prediction",
>     "problem_type": "PROBLEM_TYPE_REGRESSION",
>     "timestamp_col": "timestamp"
>   },
>   "output_schema_name": "sandbox$tdeaabhaalaji.things$tabikdcljiild",
>   "snapshot": {}
> }
< HTTP/2.0 400 Bad Request
< {
<   "error_code": "INVALID_PARAMETER_VALUE",
<   "message": "The given `info` is invalid for the following reason(s):\n - `assets_dir` is required but missing... (154 more bytes)"
< }: tf_req_id=e6f4373c-29d5-83ef-c7fd-ea412a3c369d tf_provider_addr=registry.terraform.io/hashicorp/databricks tf_rpc=ApplyResourceChange tf_resource_type=databricks_lakehouse_monitor
Error: -24T11:02:07.276Z [ERROR] sdk.proto: Response contains error diagnostic: diagnostic_detail="" tf_req_id=e6f4373c-29d5-83ef-c7fd-ea412a3c369d tf_provider_addr=registry.terraform.io/hashicorp/databricks tf_proto_version=5.4 tf_rpc=ApplyResourceChange tf_resource_type=databricks_lakehouse_monitor diagnostic_severity=ERROR
  diagnostic_summary=
  | cannot update lakehouse monitor: The given `info` is invalid for the following reason(s):
  |  - `assets_dir` is required but missing.
  |  - `table_name` is required but missing.
  |  - `profile_metrics_table_name` is required but missing.
  |  - `drift_metrics_table_name` is required but missing.
```

File: catalog/resource_lakehouse_monitor.go
Comment (mgyucht) [reply catalog/resource_lakehouse_monitor.go:25]: +1, let's not have defaults in the provider that aren't already in the API. If we want this behavior, it should be consistent with the SDKs and other API users. Could we change the REST API itself to provide a default value if not specified (i.e. make this a computed field)?

File: catalog/resource_lakehouse_monitor.go
Comment (mgyucht) [reply catalog/resource_lakehouse_monitor.go:13]: +1 for matching the path parameter exactly. `table_name` is also kind of confusing because it might suggest that you only need the table name itself for this request, but I think you want the fully qualified name including catalog/schema, right?

Hm... @alexott we should probably handle this automatically in `StructToSchema`, right? Do you know why we ignore fields that are missing the `json` tag or have `json:"-"`?

File: catalog/resource_lakehouse_monitor_test.go
Comment (mgyucht) [inline catalog/resource_lakehouse_monitor_test.go:47]: This is duplicate? Why do we need to call CreateMonitor twice? Maybe this is a bug in the unit-testing framework?

**Review by mgyucht — COMMENTED** [review]
+1 on @alexott's comments. Otherwise, this is a much nicer approach for us to maintain.

@edwardfeng-db it might be good to look at this w.r.t. my comment on StructToSchema.

File: catalog/resource_lakehouse_monitor.go
Comment (aravind-segu) [reply catalog/resource_lakehouse_monitor.go:13]: I think its because this field is part of the query  parameters in the URL rather than the request body

Comment (mgyucht) [reply catalog/resource_lakehouse_monitor.go:13]: Yes, but this structure is not used for sending requests, only for computing the TF schema.

File: catalog/resource_lakehouse_monitor_test.go
Comment (aravind-segu) [reply catalog/resource_lakehouse_monitor_test.go]: removed it.

**Comment by alexott** [comment]
Integration test is still failing:

```
    init_test.go:236: Step 1/2 error: After applying this test step, the plan was not empty.
        stdout:
        
        
        Terraform used the selected providers to generate the following execution
        plan. Resource actions are indicated with the following symbols:
          ~ update in-place
        
        Terraform will perform the following actions:
        
          # databricks_lakehouse_monitor.testMonitorInference will be updated in-place
          ~ resource "databricks_lakehouse_monitor" "testMonitorInference" {
              - drift_metrics_table_name   = "sandbox$tdkkfakkkfadj.things$tdkkfakkkfadj.bar$tdkkfakkkfadj_inference_drift_metrics" -> null
                id                         = "sandbox$tdkkfakkkfadj.things$tdkkfakkkfadj.bar$tdkkfakkkfadj_inference"
              - monitor_version            = "0" -> null
              - profile_metrics_table_name = "sandbox$tdkkfakkkfadj.things$tdkkfakkkfadj.bar$tdkkfakkkfadj_inference_profile_metrics" -> null
              - status                     = "MONITOR_STATUS_PENDING" -> null
                # (3 unchanged attributes hidden)
        
                # (1 unchanged block hidden)
            }
        
        Plan: 0 to add, 1 to change, 0 to destroy.
```

**Comment by alexott** [comment]
after fixing computed fields, another error in update  test:

```
    init_test.go:236: Step 2/2 error: Error running apply: exit status 1
        
        Error: cannot update lakehouse monitor: Data Monitor 'sandbox$tijgeledgkhhb.things$tejcfjihaeblg.bar$tejcfjihaeblg_inference' does not exist.
        
          with databricks_lakehouse_monitor.testMonitorInference,
          on terraform_plugin_test.tf line 36, in resource "databricks_lakehouse_monitor" "testMonitorInference":
          36:         resource "databricks_lakehouse_monitor" "testMonitorInference" {
        

```

**Comment by aravind-segu** [comment]
Integration tests are passing: https://github.com/databricks-eng/eng-dev-ecosystem/actions/runs/8147735527

**Review by alexott — CHANGES_REQUESTED** [review]
Small changes are still required in the code. 

Plus there is no documentation yet

File: catalog/resource_lakehouse_monitor.go
Comment (mgyucht) [inline catalog/resource_lakehouse_monitor.go:19]: The SDK supports long-running operations automatically through the `x-databricks-wait` annotation. Will send you the link offline. This will let you use the following standard pattern for long-running operations:
```go
wait, err := w.LakehouseMonitors.Create(ctx, req)
if err != nil {
  return err
}
wait.Get()
```

File: catalog/resource_lakehouse_monitor.go
Comment (mgyucht) [inline catalog/resource_lakehouse_monitor.go:39]: These can be removed, based on the specification: these fields are not marked `omitempty`, so they will be required in the Terraform schema.

File: catalog/resource_lakehouse_monitor.go
Comment (mgyucht) [inline catalog/resource_lakehouse_monitor.go:40]: This does not seem to match any field in the create request. What field does this correspond to? I think this should be `SetReadOnly()`.

File: catalog/resource_lakehouse_monitor.go
Comment (mgyucht) [inline catalog/resource_lakehouse_monitor.go:69]: Isn't this name part of the Create response? How does the user know this ahead of time?

**Review by mgyucht — COMMENTED** [review]
Two main comments:
1. Waiters are supported in the SDK by default, you just need to annotate your API appropriately. I'll send you the link to this offline.
2. How exactly is table_name supposed to work? It seems like it is a required field but it is not in the Create or Update requests.

File: docs/resources/lakehouse_monitor.md
Comment (alexott) [inline docs/resources/lakehouse_monitor.md:122]: We should document the computed fields: `monitor_version`, `drift_metrics_table_name`, `profile_metrics_table_name`, `status`, `dashboard_id`

**Review by alexott — COMMENTED** [review]
In general looks good - we need to decide about readiness waiting - should we wait for new Go SDK, or implement it later

File: catalog/resource_lakehouse_monitor.go
Comment (aravind-segu) [reply catalog/resource_lakehouse_monitor.go:39]: The struct we are using here is the `MonitorInfo` and not `CreateMonitor`. `MonitorInfo` has additional fields such as `drift_metrics_table_name`, `profile_metrics_table_name` etc that are not present in `CreateSchema`. Therefore if we use `CreateMonitor`, the GET request is not translating all the fields correctly. 

In the `MonitorInfo` struct, assets_dir, output_schema_name and table_name are not set to `omitempty`. So this forces them to be required.

File: catalog/resource_lakehouse_monitor.go
Comment (aravind-segu) [reply catalog/resource_lakehouse_monitor.go:40]: We are using `MonitorInfo` as the base struct and this is part of that struct

File: catalog/resource_lakehouse_monitor.go
Comment (aravind-segu) [reply catalog/resource_lakehouse_monitor.go:69]: Yes this is the `FullName` which is a path parameter. The name here is the same as the table name that the user wants to attach the monitor to. Therefore they do know this ahead of time.

File: docs/resources/lakehouse_monitor.md
Comment (alexott) [inline docs/resources/lakehouse_monitor.md:123]: small nit - it's better to move it to the `Attribute Reference` section

**Review by alexott — APPROVED** [review]
in general good, pending decision on if we should wait for OpenAPI spec changes for wait command

**Comment by aravind-segu** [comment]
> in general good, pending decision on if we should wait for OpenAPI spec changes for wait command

I checked with Miles, and he is ok to push this for now. I will wait for his approval as well

File: catalog/resource_lakehouse_monitor.go
Comment (mgyucht) [reply catalog/resource_lakehouse_monitor.go:39]: Thanks for clarifying. This is fine then.

File: catalog/resource_lakehouse_monitor.go
Comment (mgyucht) [reply catalog/resource_lakehouse_monitor.go:40]: Alright. Sounds reasonable to me.

File: docs/resources/lakehouse_monitor.md
Comment (mgyucht) [inline docs/resources/lakehouse_monitor.md:13]:
```suggestion
```

File: docs/resources/lakehouse_monitor.md
Comment (mgyucht) [inline docs/resources/lakehouse_monitor.md:31]:
```suggestion
resource "databricks_sql_table" "myTestTable" {
```

File: docs/resources/lakehouse_monitor.md
Comment (mgyucht) [inline docs/resources/lakehouse_monitor.md:43]: Is this right? I think this has extra information.
```suggestion
        type = "int"
```

**Review by mgyucht — APPROVED** [review]
LGTM, provided you address @alexott's comment to move computed fields under the Attribute Reference section of the doc.

File: docs/resources/lakehouse_monitor.md
Comment (arpitjasa-db) [inline docs/resources/lakehouse_monitor.md:8]: It can be attached to any UC table right, not just this specific SQL table which is a workaround to initialize tables via Terraform? Maybe we should call out that it can be used with the table data source as well

File: docs/resources/lakehouse_monitor.md
Comment (arpitjasa-db) [inline docs/resources/lakehouse_monitor.md:48]: Hmm I wonder if this is a bit too much, we ended up separating this in a UC Model into 3 parameters for this reason: https://registry.terraform.io/providers/databricks/databricks/latest/docs/resources/registered_model

File: docs/resources/lakehouse_monitor.md
Comment (arpitjasa-db) [inline docs/resources/lakehouse_monitor.md:80]: Looks like a bit weird syntax, would it be better to expose this as a boolean instead and then convert in the resource implementation?

File: docs/resources/lakehouse_monitor.md
Comment (arpitjasa-db) [inline docs/resources/lakehouse_monitor.md:90]: Let's clarify that it needs to be catalog.schema

File: docs/resources/lakehouse_monitor.md
Comment (alexott) [inline docs/resources/lakehouse_monitor.md:45]: This could be simplified to `databricks_sql_table.myTestTable.id` now: https://registry.terraform.io/providers/databricks/databricks/latest/docs/resources/sql_table#id

**Review by alexott — APPROVED** [review]

File: docs/resources/lakehouse_monitor.md
Comment (aravind-segu) [reply docs/resources/lakehouse_monitor.md]: According to their python api docs (https://api-docs.databricks.com/python/lakehouse-monitoring/latest/databricks.lakehouse_monitoring.html#databricks.lakehouse_monitoring.create_monitor) , it can be `{catalog}.{schema}.{table}` or `{schema}.{table}` or `{table}` and the api fills in the current catalog or schema. So we technically dont need to split it up and force users to fill in all three fields. I think this should be the recommendation to fill in all three, but if the user reads other docs, they should be able to just use all three of the options

Comment (arpitjasa-db) [reply docs/resources/lakehouse_monitor.md]: Got it, so if someone doesn't specify schema and catalog via Terraform though, what would be the "current" catalog and schema it would infer?

File: docs/resources/lakehouse_monitor.md
Comment (aravind-segu) [reply docs/resources/lakehouse_monitor.md:77]: We debated about this. We changed the Go SDK to use an empty struct in place of `any`. This is also the the example expected in the Go SDK, so thought it would be clear. Miles also did not want too many changes between Go SDK structs and the terraform input structs. 

Also in the future if there are any snapshot relevant parameters the team introduces, we dont need additional changes in the terraform provider as we are already using the struct.

File: docs/resources/lakehouse_monitor.md
Comment (alexott) [reply docs/resources/lakehouse_monitor.md]: No, in for `databricks_sql_table` you always specify catalog & schema and `id` will be three-level name

**Comment by samuhepp** [comment]
Hey! Will this be included in the next release? If so, when is the approximate target date for that? Keen to use this instead of the notebook approach.

**Pull request merged**
